nats/nats.go

```diff
@@ -0,0 +1,368 @@
+package nats
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/nats.go/jetstream"
+)
+
+// Storage interface that is implemented by storage providers
+type Storage struct {
+	nc  *nats.Conn
+	kv  jetstream.KeyValue
+	err []error
+	ctx context.Context
+	cfg Config
+	mu  sync.RWMutex
+}
+
+type entry struct {
+	Data   []byte
+	Expiry int64
+}
+
+func init() {
+	gob.Register(entry{})
+}
+
+// logErrorw is a helper function to log error messages
+func (s *Storage) logErrorw(msg string, keysAndValues ...interface{}) {
+	if s.cfg.UseLogger && s.cfg.Logger != nil {
+		s.cfg.Logger.Errorw(msg, keysAndValues...)
+	}
+}
+
+// logInfow is a helper function to log error messages
+func (s *Storage) logInfow(msg string, keysAndValues ...interface{}) {
+	if s.cfg.UseLogger && s.cfg.Logger != nil {
+		s.cfg.Logger.Infow(msg, keysAndValues...)
+	}
+}
+
+// connectHandler is a helper function to set the initial connect handler
+func (s *Storage) connectHandler(nc *nats.Conn) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.logInfow("connected",
+		"diver", "nats",
+		"url", nc.ConnectedUrlRedacted(),
+	)
+
+	var err error
+	s.kv, err = newNatsKV(
+		nc,
+		s.ctx,
+		s.cfg.KeyValueConfig,
+	)
+	if err != nil {
+		s.logErrorw("kv not initialized",
+			"diver", "nats",
+			"error", err.Error(),
+		)
+		s.err = append(s.err, err)
+	}
+}
+
+// disconnectErrHandler is a helper function to set the disconnect error handler
+func (s *Storage) disconnectErrHandler(nc *nats.Conn, err error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if err != nil {
+		s.logErrorw("disconnected",
+			"diver", "nats",
+			"error", err.Error(),
+		)
+	} else {
+		s.logInfow("disconnected",
+			"diver", "nats",
+		)
+	}
+	nc.Opts.RetryOnFailedConnect = true
+	if err != nil {
+		s.err = append(s.err, err)
+	}
+}
+
+// reconnectHandler is a helper function to set the reconnect handler
+func (s *Storage) reconnectHandler(nc *nats.Conn) {
+	s.connectHandler(nc)
+}
+
+// errorHandler is a helper function to set the error handler
+func (s *Storage) errorHandler(nc *nats.Conn, sub *nats.Subscription, err error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	s.logErrorw("error handler",
+		"diver", "nats",
+		"sub", sub.Subject,
+		"error", err.Error(),
+	)
+	if err != nil {
+		s.err = append(s.err, fmt.Errorf("subject %q: %w", sub.Subject, err))
+	}
+}
+
+// closedHandler is a helper function to set the closed handler
+func (s *Storage) closedHandler(nc *nats.Conn) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
```

Comment:
Why get a lock here to log a message? This will reduce performance with logs enable.